File: python/vierrataleai/providers/anthropic.py
import json
import urllib.request
import urllib.error
from typing import AsyncGenerator, List, Optional

from .base import BaseProvider
from ..catalog import get_real_model, get_display_name, get_all_models
from .. import config


class AnthropicProvider(BaseProvider):
    def __init__(self):
        super().__init__("anthropic")

    @property
    def display_name(self):
        return "Nebula"

    def _resolve_model(self, model_name: str) -> str:
        if model_name.startswith("claude-"):
            return model_name
        return "claude-3-5-haiku-20241022"

    async def is_available(self) -> bool:
        return bool(config.get("anthropic_api_key"))

    async def list_models(self) -> List[dict]:
        if not config.get("anthropic_api_key"):
            return []
        return [
            {"real_name": real, "display_name": get_display_name(real)}
            for real in get_all_models()
            if real.startswith("claude-")
        ]

    async def stream(
        self,
        messages: List[dict],
        options: Optional[dict] = None,
    ) -> AsyncGenerator[str, None]:
        options = options or {}
        key = config.get("anthropic_api_key")
        if not key:
            raise RuntimeError("[ERR-0004] Cloud API key not configured. Add it to the config and try again.")

        model = self._resolve_model(options.get("model", config.get("model")))
        system_prompt = options.get("system_prompt", "")

        system = []
        if system_prompt:
            system.append({"type": "text", "text": system_prompt})
        api_messages = [
            {"role": m["role"], "content": m["content"]}
            for m in messages
        ]

        body = json.dumps({
            "model": model,
            "system": system if system else None,
            "messages": api_messages,
            "max_tokens": options.get("max_tokens", config.get("max_tokens")),
            "stream": True,
            "temperature": options.get("temperature", config.get("temperature")),
        }).encode()

        req = urllib.request.Request(
            "https://api.anthropic.com/v1/messages",
            data=body,
            headers={
                "Content-Type": "application/json",
                "x-api-key": key,
                "anthropic-version": "2023-06-01",
            },
        )

        try:
            resp = urllib.request.urlopen(req, timeout=240)
        except urllib.error.HTTPError as err:
            raise RuntimeError(
                f"[ERR-0003] The cloud provider returned HTTP {err.code}. Check your API key in config."
            )
        except Exception as err:
            raise RuntimeError("[ERR-0001] Could not reach the API (no response within 120s). Check your network connection.")

        buffer = ""
        for chunk in iter(lambda: resp.read(1024), b""):
            buffer += chunk.decode("utf-8", errors="ignore")
            lines = buffer.split("\n")
            buffer = lines.pop() or ""
            for line in lines:
                if not line.startswith("data: "):
                    continue
                data_str = line[6:]
                try:
                    data = json.loads(data_str)
                    if data.get("type") == "content_block_delta":
                        text = data.get("delta", {}).get("text", "")
                        if text:
                            yield text
                except json.JSONDecodeError:
                    continue

File: nodejs/src/providers/anthropic.js
import { BaseProvider } from './base.js';
import { Catalog } from '../catalog.js';
import { Config } from '../config.js';

const ANTHROPIC_URL = 'https://api.anthropic.com/v1/messages';

export class AnthropicProvider extends BaseProvider {
  constructor() {
    super('anthropic');
  }

  get displayName() {
    return 'Nebula';
  }

  _resolveModel(options) {
    const requested = options.model || Config.get('model');
    if (Catalog.isLocalModel(requested) || !Catalog.isCloudModel(requested)) {
      return Catalog.getRealModel(Catalog.getDefaultCloudModel());
    }
    const real = Catalog.getRealModel(requested);
    return real.startsWith('claude-') ? real : 'claude-3-5-haiku-20241022';
  }

  async isAvailable() {
    const key = Config.get('anthropicApiKey');
    return Boolean(key);
  }

  async listModels() {
    const key = Config.get('anthropicApiKey');
    if (!key) return [];
    const models = Object.keys(Catalog.getAllModels()).filter((m) => m.startsWith('claude-'));
    return models.map((real) => ({
      realName: real,
      displayName: Catalog.getDisplayName(real),
    }));
  }

  async *stream(messages, options = {}) {
    const key = Config.get('anthropicApiKey');
    if (!key) throw new Error('[ERR-0004] Cloud API key not configured. Add it to the config and try again.');

    const model = this._resolveModel(options);
    const systemPrompt = options.systemPrompt || '';

    const system = [];
    if (systemPrompt) system.push({ type: 'text', text: systemPrompt });
    const apiMessages = messages.map((m) => ({
      role: m.role === 'assistant' ? 'assistant' : 'user',
      content: m.content,
    }));

    const controller = new AbortController();
    const connectTimer = setTimeout(() => controller.abort(), 240000);

    let resp;
    try {
      resp = await fetch(ANTHROPIC_URL, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': key,
          'anthropic-version': '2023-06-01',
        },
        body: JSON.stringify({
          model,
          system: system.length ? system : undefined,
          messages: apiMessages,
          max_tokens: options.maxTokens || Config.get('maxTokens'),
          stream: true,
          temperature: options.temperature || Config.get('temperature'),
        }),
        signal: controller.signal,
      });
    } catch (err) {
      if (err.name === 'AbortError') {
        throw new Error('[ERR-0001] Could not reach the API (no response within 60s). Check your network connection.');
      }
      throw new Error('[ERR-0001] Could not reach the API (no response). Check your network connection.');
    } finally {
      clearTimeout(connectTimer);
    }

    if (!resp.ok) {
      throw new Error('[ERR-0003] The cloud provider returned HTTP ' + resp.status + '. Check your API key in config.');
    }

    const reader = resp.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    let idle = null;
    const settle = () => {
      if (idle) clearTimeout(idle);
      idle = setTimeout(() => controller.abort(), 240000);
    };
    settle();
    try {
      while (true) {
        let value, done;
        try {
          ({ done, value } = await reader.read());
        } catch (err) {
          if (err.name === 'AbortError') {
            throw new Error('[ERR-0001] Lost the connection to the cloud API (no data for 60s).');
          }
          throw err;
        }
        if (done) break;
        settle();
        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() || '';

        for (const line of lines) {
          if (!line.startsWith('data: ')) continue;
          const data = line.slice(6);
          try {
            const json = JSON.parse(data);
            if (json.type === 'content_block_delta' && json.delta?.text) {
              yield json.delta.text;
            }
          } catch {}
        }
      }
    } finally {
      if (idle) clearTimeout(idle);
    }
  }
}
